File: ablySpec/NSObject+TestSuite.m
//
//  NSObject+TestSuite.m
//  ably
//
//  Created by Ricardo Pereira on 08/02/16.
//  Copyright © 2016 Ably. All rights reserved.
//

#import "NSObject+TestSuite.h"
#import <Aspects/Aspects.h>

@implementation NSObject (TestSuite)

- (void)testSuite_getReturnValueFrom:(SEL)selector callback:(void (^)(id))callback {
    [self aspect_hookSelector:selector withOptions:AspectPositionAfter usingBlock:^(id<AspectInfo> info) {
        __autoreleasing id result;
        [[info originalInvocation] getReturnValue:&result];
        callback([result copy]);
    } error:nil];
}

- (void)testSuite_injectIntoMethod:(SEL)selector code:(void (^)(void))block {
    [self aspect_hookSelector:selector withOptions:AspectPositionAfter usingBlock:^(id<AspectInfo> info) {
        block();
    } error:nil];
}

@end

File: ablySpec/NSObject+TestSuite.swift
//
//  NSObject+TestSuite.swift
//  ably
//
//  Created by Ricardo Pereira on 17/02/16.
//  Copyright © 2016 Ably. All rights reserved.
//

import Aspects

extension NSObject {

    /// Inject a block of code to the identified class method.
    class func testSuite_injectIntoClassMethod(selector: Selector, code: ()->()) {
        let block: @convention(block) (AspectInfo) -> Void = { _ in
            code()
        }
        let _ = try! self.aspect_hookSelector(selector, withOptions: .PositionAfter, usingBlock: unsafeBitCast(block, AnyObject.self))
    }

}
